File: top20_biggest.py
#!/usr/bin/env python3
import os
from pathlib import Path
import heapq
import pathspec

# 1. Load .gitignore
gitignore_file = Path(".gitignore")
if gitignore_file.exists():
    with gitignore_file.open("r") as f:
        gitignore_patterns = f.read().splitlines()
else:
    gitignore_patterns = []

spec = pathspec.PathSpec.from_lines("gitwildmatch", gitignore_patterns)

# 2. Traverse the directory
file_sizes = []

for root, dirs, files in os.walk("."):
    # Remove ignored directories from traversal
    dirs[:] = [d for d in dirs if not spec.match_file(os.path.relpath(os.path.join(root, d), "."))]
    for file in files:
        filepath = os.path.join(root, file)
        relpath = os.path.relpath(filepath, ".")
        if spec.match_file(relpath):
            continue
        try:
            size = os.path.getsize(filepath)
            file_sizes.append((size, relpath))
        except OSError:
            continue

# 3. Get top 20 largest files
top20 = heapq.nlargest(20, file_sizes)

print("Top 20 largest files (size in bytes):")
for size, path in top20:
    print(f"{size:>10}  {path}")

File: top20_biggest.rb
#!/usr/bin/env ruby
require 'pathname'

# Load .gitignore patterns
gitignore_file = Pathname.new(".gitignore")
patterns = []
if gitignore_file.exist?
  gitignore_file.read.each_line do |line|
    line = line.strip
    next if line.empty? || line.start_with?("#")
    patterns << line
  end
end

# Check if a path is ignored according to patterns
def ignored?(path, patterns)
  path = path.gsub("\\", "/") # normalize
  ignore = false
  patterns.each do |pattern|
    negated = pattern.start_with?("!")
    pat = negated ? pattern[1..] : pattern

    # Convert gitignore pattern to glob
    glob = pat
    glob = glob.sub(%r{^\./}, '') # remove leading ./
    
    # Use File.fnmatch with FNM_PATHNAME
    if File.fnmatch?(glob, path, File::FNM_PATHNAME | File::FNM_DOTMATCH)
      ignore = !negated
    end
  end
  ignore
end

# Traverse all files recursively
file_sizes = []
Dir.glob("**/*", File::FNM_DOTMATCH).each do |f|
  next if File.directory?(f)
  next if ignored?(f, patterns)
  begin
    file_sizes << [File.size(f), f]
  rescue
    next
  end
end

# Top 20 largest
top20 = file_sizes.sort_by { |size, _| -size }.first(20)

puts "Top 20 largest files (size in bytes):"
top20.each do |size, path|
  puts "%10d  %s" % [size, path]
end

